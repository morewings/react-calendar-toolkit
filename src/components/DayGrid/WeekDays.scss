@import "./../../variables";

.weekdays-wrapper{
  display: flex;
}