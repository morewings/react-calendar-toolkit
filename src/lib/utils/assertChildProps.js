import React, {Fragment} from 'react';

export const trackProps = jest.fn();

export const DummyComponent = props => {
  trackProps(props);
  return <Fragment>Child</Fragment>;
};
