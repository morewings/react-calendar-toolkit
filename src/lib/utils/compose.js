const compose = (...fns) => (...args) =>
  fns.reduceRight(
    (params, f) => (Array.isArray(params) ? f(...params) : f(params)),
    args
  );

export default compose;
