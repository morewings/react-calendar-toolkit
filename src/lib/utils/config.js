export default Object.freeze({
  supportedPrecisions: Object.freeze([
    'year',
    'month',
    'day',
    'hour',
    'minute',
  ]),
});
