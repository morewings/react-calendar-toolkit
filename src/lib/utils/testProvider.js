import React from 'react';
import PropTypes from 'prop-types';
import {Provider as LocaleProvider} from 'lib/features/locale/withLocaleContext';
import {Provider as ThemeProvider} from 'lib/features/theme/withTheme';
import {Provider as ModalProvider} from 'lib/features/modal/withModalProvider';
import {Provider as DatepickerProvider} from 'lib/features/datepicker/withDatepickerProvider';

export const DatepickerMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <DatepickerProvider>{props.children}</DatepickerProvider>
    </ThemeProvider>
  </LocaleProvider>
);

export const InputMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <ModalProvider>
        <DatepickerProvider>{props.children}</DatepickerProvider>
      </ModalProvider>
    </ThemeProvider>
  </LocaleProvider>
);

DatepickerMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

InputMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
