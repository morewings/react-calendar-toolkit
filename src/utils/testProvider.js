import React from 'react';
import PropTypes from 'prop-types';
import {Provider as LocaleProvider} from 'features/locale/withLocaleContext';
import {Provider as ThemeProvider} from 'features/theme/withTheme';
import {Provider as ModalProvider} from 'features/modal/withModalProvider';
import {Provider as DatepickerProvider} from 'features/datepicker/withDatepickerProvider';

export const DatepickerMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <DatepickerProvider>{props.children}</DatepickerProvider>
    </ThemeProvider>
  </LocaleProvider>
);

export const InputMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <ModalProvider>
        <DatepickerProvider>{props.children}</DatepickerProvider>
      </ModalProvider>
    </ThemeProvider>
  </LocaleProvider>
);

DatepickerMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

InputMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
