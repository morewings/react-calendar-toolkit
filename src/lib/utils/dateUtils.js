import {
  formatWithOptions,
  isSameMonth,
  isSameYear,
  toDate,
  getTime,
  addMonths,
  addDays,
  addYears,
  isWithinInterval,
  isWeekend,
  isSameDay,
  isSameHour,
  getDate,
  startOfMonth,
  endOfMonth,
} from 'date-fns/fp';
import compose from 'lib/utils/compose';

/**
 * Unix timestamp
 * @typedef {number} UnixTimestamp
 */

/**
 * Date union type may be Date or Unix timestamp
 * @typedef {(Date|UnixTimestamp)} DateUnion
 */

/**
 * @function
 * @name formatDateWithLocale
 * @description Returns formatted date
 * @param {Object} options - Options object with locale
 * @param {string} pattern - Formatting pattern
 * @param {Date} date - Date object to apply format
 * @return {string} Formatted date
 */
export const formatDateWithLocale = formatWithOptions;

/**
 * @function
 * @name checkIsSameMonth
 * @description Checks if two dates belong to the same month
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameMonth = isSameMonth;

/**
 * @function
 * @name checkIsSameYear
 * @description Checks if two dates belong to the same year
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameYear = isSameYear;

/**
 * @function
 * @name checkIsSameDay
 * @description Checks if two dates belong to the same day
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameDay = isSameDay;

/**
 * @function
 * @name checkIsSameHour
 * @description Checks if two dates belong to the same hour
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameHour = isSameHour;

/**
 * @function
 * @name convertToTimestamp
 * @description Converts provided date to Unix timestamp
 * @param {Date} date - Date to convert
 * @return {UnixTimestamp} Unix timestamp
 */
export const convertToTimestamp = date => getTime(date);

/**
 * @function
 * @name convertToDate
 * @description Converts provided Unix timestamp to Date object
 * @param {number} timestamp - Date to convert
 * @return {Date}
 */
export const convertToDate = timestamp => toDate(timestamp);

/**
 * @function
 * @name convertToDayOfMonth
 * @description Converts provided Unix timestamp or Date to day of month number
 * @param {DateUnion} timestamp - Timestamp or Date to convert
 * @return {number} Day of the month
 */
export const convertToDayOfMonth = timestamp => getDate(timestamp);

/**
 * @function
 * @name checkIsWithinInterval
 * @description Checks if date belongs to the interval
 * @typedef {Object} Interval
 * @property {Date} start - Start of interval
 * @property {Date} end - End of interval
 * @param {DateUnion} date - Date to check
 * @param {Interval} interval - Time range to check
 * @return {boolean}
 */
export const checkIsWithinInterval = isWithinInterval;

/**
 * @function
 * @name checkIsWeekend
 * @description Checks if date belongs to the weekend
 * @param {DateUnion} date - Date to check
 * @return {boolean}
 */
export const checkIsWeekend = date => isWeekend(date);

/**
 * @function
 * @name matchDatesWithPrecision
 * @description Checks if dates are the same with given precision, eg belong to same day|month|year
 * @param {string} precision - Date to check
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const matchDatesWithPrecision = (precision, dateLeft, dateRight) =>
  ({
    day: checkIsSameDay,
    month: checkIsSameMonth,
    year: checkIsSameYear,
  }[precision](dateLeft, dateRight));

/**
 * @function
 * @name ceilMonth
 * @description Gets end of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const ceilMonth = endOfMonth;

/**
 * @function
 * @name floorMonth
 * @description Gets start of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const floorMonth = startOfMonth;

/**
 * @function
 * @name addMonth
 * @description Adds months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to add
 * @return {UnixTimestamp}
 */
export const addMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(amount, date);

/**
 * @function
 * @name subMonth
 * @description Subtracts months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to subtract
 * @return {UnixTimestamp}
 */
export const subMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(-amount, date);

/**
 * @function
 * @name addDay
 * @description Adds days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to add
 * @return {UnixTimestamp}
 */
export const addDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(amount, date);

/**
 * @function
 * @name subDay
 * @description Subtracts days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(-amount, date);

/**
 * @function
 * @name addYear
 * @description Adds years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of years to add
 * @return {UnixTimestamp}
 */
export const addYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(amount, date);

/**
 * @function
 * @name subYear
 * @description Subtracts years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(-amount, date);
