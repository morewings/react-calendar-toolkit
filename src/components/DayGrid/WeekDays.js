/*eslint-disable*/
import React from 'react';
import {getWeekDays} from 'utils/dateUtils';
import './WeekDays.scss';

const WeekDays = props => {
  const weekDays = getWeekDays();
  return (
    <div className="weekdays-wrapper">
      {weekDays.map(day => (
        <div className={} key={day}>{day}</div>
      ))}
    </div>
  );
};

export default WeekDays;
