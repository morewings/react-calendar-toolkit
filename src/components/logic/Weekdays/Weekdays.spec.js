import React from 'react';
import {render} from '@testing-library/react';
import {DatepickerMockProvider as Provider} from 'utils/testProvider';
import customLocale from 'date-fns/locale/ru';
import WeekDays from './Weekdays';

const MockComponent = jest.fn(() => <div>MockComponent</div>);

const MockWrapper = jest.fn(({children}) => <div>{children}</div>);

const renderWithProviderProps = (Component, props = {}) =>
  render(<Component />, {
    wrapper: ({children}) => <Provider {...props}>{children}</Provider>,
  });

describe('WeekDays', () => {
  beforeEach(() => {
    MockComponent.mockClear();
  });

  it('renders with default locale', () => {
    const {asFragment} = renderWithProviderProps(() => (
      <WeekDays renderAs={MockComponent} wrapWith={MockWrapper} />
    ));
    expect(asFragment()).toMatchSnapshot();

    MockComponent.mock.calls.forEach(call => {
      expect(call[0]).toMatchSnapshot();
    });
  });

  it('renders with custom locale', () => {
    renderWithProviderProps(
      () => <WeekDays renderAs={MockComponent} wrapWith={MockWrapper} />,
      {dateFnsLocale: customLocale}
    );

    MockComponent.mock.calls.forEach(call => {
      expect(call[0]).toMatchSnapshot();
    });
  });
});
