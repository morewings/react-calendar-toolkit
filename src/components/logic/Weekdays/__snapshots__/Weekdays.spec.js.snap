// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeekDays renders with custom locale 1`] = `
Object {
  "name": Object {
    "abbreviated": "пнд",
    "narrow": "П",
    "numeric": 0,
    "short": "пн",
    "wide": "понедельник",
  },
}
`;

exports[`WeekDays renders with custom locale 2`] = `
Object {
  "name": Object {
    "abbreviated": "втр",
    "narrow": "В",
    "numeric": 1,
    "short": "вт",
    "wide": "вторник",
  },
}
`;

exports[`WeekDays renders with custom locale 3`] = `
Object {
  "name": Object {
    "abbreviated": "срд",
    "narrow": "С",
    "numeric": 2,
    "short": "ср",
    "wide": "среда",
  },
}
`;

exports[`WeekDays renders with custom locale 4`] = `
Object {
  "name": Object {
    "abbreviated": "чтв",
    "narrow": "Ч",
    "numeric": 3,
    "short": "чт",
    "wide": "четверг",
  },
}
`;

exports[`WeekDays renders with custom locale 5`] = `
Object {
  "name": Object {
    "abbreviated": "птн",
    "narrow": "П",
    "numeric": 4,
    "short": "пт",
    "wide": "пятница",
  },
}
`;

exports[`WeekDays renders with custom locale 6`] = `
Object {
  "name": Object {
    "abbreviated": "суб",
    "narrow": "С",
    "numeric": 5,
    "short": "сб",
    "wide": "суббота",
  },
}
`;

exports[`WeekDays renders with custom locale 7`] = `
Object {
  "name": Object {
    "abbreviated": "вск",
    "narrow": "В",
    "numeric": 6,
    "short": "вс",
    "wide": "воскресенье",
  },
}
`;

exports[`WeekDays renders with default locale 1`] = `
<DocumentFragment>
  <div>
    <div>
      MockComponent
    </div>
    <div>
      MockComponent
    </div>
    <div>
      MockComponent
    </div>
    <div>
      MockComponent
    </div>
    <div>
      MockComponent
    </div>
    <div>
      MockComponent
    </div>
    <div>
      MockComponent
    </div>
  </div>
</DocumentFragment>
`;

exports[`WeekDays renders with default locale 2`] = `
Object {
  "name": Object {
    "abbreviated": "Sun",
    "narrow": "S",
    "numeric": 0,
    "short": "Su",
    "wide": "Sunday",
  },
}
`;

exports[`WeekDays renders with default locale 3`] = `
Object {
  "name": Object {
    "abbreviated": "Mon",
    "narrow": "M",
    "numeric": 1,
    "short": "Mo",
    "wide": "Monday",
  },
}
`;

exports[`WeekDays renders with default locale 4`] = `
Object {
  "name": Object {
    "abbreviated": "Tue",
    "narrow": "T",
    "numeric": 2,
    "short": "Tu",
    "wide": "Tuesday",
  },
}
`;

exports[`WeekDays renders with default locale 5`] = `
Object {
  "name": Object {
    "abbreviated": "Wed",
    "narrow": "W",
    "numeric": 3,
    "short": "We",
    "wide": "Wednesday",
  },
}
`;

exports[`WeekDays renders with default locale 6`] = `
Object {
  "name": Object {
    "abbreviated": "Thu",
    "narrow": "T",
    "numeric": 4,
    "short": "Th",
    "wide": "Thursday",
  },
}
`;

exports[`WeekDays renders with default locale 7`] = `
Object {
  "name": Object {
    "abbreviated": "Fri",
    "narrow": "F",
    "numeric": 5,
    "short": "Fr",
    "wide": "Friday",
  },
}
`;

exports[`WeekDays renders with default locale 8`] = `
Object {
  "name": Object {
    "abbreviated": "Sat",
    "narrow": "S",
    "numeric": 6,
    "short": "Sa",
    "wide": "Saturday",
  },
}
`;
